### PR TITLE
Shifted TokenActivity to use a ViewModel and switched its usage of ProgressDialog

### PR DESCRIPTION
**Pull request by aryanarora180**
All of TokenActivity's business logic has been shifted to a ViewModel. Its usage of AsyncTask has been shifted to use Coroutines, LiveData, and Observers. Moreover, the usage of deprecated ProgressDialog has also been switched to a simple ProgressBar and accompanying text on the screen.
Login screen 
Login screen - fetching data

### CODE REVIEW

File: app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        data["SITE_URL"] = Urls.MOODLE_URL.toString().replace("/$".toRegex(), "")
+        data["PASSPORT"] = passport
+        MyApplication.instance.setLoginLaunchData(data)
+        val intent = Intent(Intent.ACTION_VIEW)
+        intent.data = Uri.parse(loginUrl)
+        _startIntent.postValue(SingleLiveEvent(intent))
+    }
+
+    companion object {
+        const val FETCHING_DETAILS = 0
```
Comment (abhijeetviswa) [inline app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt:217]: Looks like this value is no longer being used. You can get rid of it and renumber the other status

File: app/src/main/java/crux/bphc/cms/activities/TokenActivity.kt [hunk 2]
```diff
@@ -45,232 +28,74 @@ class TokenActivity : AppCompatActivity() {
...
 
-    private fun showProgress(show: Boolean, message: String) {
-        if (show) progressDialog.show() else progressDialog.hide()
-        progressDialog.setMessage(message)
-    }
-
-    private fun dismissProgress() {
-        if (progressDialog.isShowing) {
-            progressDialog.dismiss()
+    private val onErrorMessageObserver = Observer<SingleLiveEvent<String>> { error ->
```
Comment (abhijeetviswa) [inline app/src/main/java/crux/bphc/cms/activities/TokenActivity.kt:43]: I prefer putting instance var and vals above instance methods. Not to mention, these are being referenced in `onCreate`. Lmk what you feel about making this change.

File: app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+import java.security.NoSuchAlgorithmException
+import java.util.*
+
+class TokenViewModel(application: Application) : AndroidViewModel(application) {
+
+    private val TAG = javaClass.simpleName
+
+    private val moodleServices: MoodleServices = APIClient.getRetrofitInstance().create(MoodleServices::class.java)
+
+    private val _onErrorMessage: MutableLiveData<SingleLiveEvent<String>> = MutableLiveData()
```
Comment (abhijeetviswa) [inline app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt:43]: Do we really need to prepend private variables with an `_`?

**Comment by abhijeetviswa** [comment]
This looks good. A change for the better. 
I've left some comments and reviews. Also, update the changelog and amend `b1b4172c9455e9c7c99d272223002d4be14115e5`. The commit subject can also rewritten to something like: "_Replace TokenActivity ProgressDialog with on screen prompts_".

File: app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt (hunk 1 above)
Comment (aryanarora180) [reply app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt:217]: It is being used when the user is redirected to the browser to sign in with Google though right?

File: app/src/main/java/crux/bphc/cms/activities/TokenActivity.kt (hunk 2 above)
Comment (aryanarora180) [reply app/src/main/java/crux/bphc/cms/activities/TokenActivity.kt:43]: For Observers, it's convention to keep them below onCreate since they manipulate the UI, however, all other instance vals and vars should be kept above the instance methods.

File: app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt (hunk 3 above)
Comment (aryanarora180) [reply app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt:43]: I am a little confused myself about prepending private variables with an `_` too, but again, from the convention I've seen, the LiveData is always the MutableLiveData's variable name prepended with an `_`. Since only the ViewModel changes the value of the MutableLiveData and the Fragment should only be able to read LiveData, MutableLiveData is private and has the `_` before its variable name.

File: app/src/main/java/crux/bphc/cms/activities/TokenActivity.kt (hunk 2 above)
Comment (abhijeetviswa) [reply app/src/main/java/crux/bphc/cms/activities/TokenActivity.kt:43]: Alright, as long it's an established convention.

File: CHANGELOG.md [hunk 4]
```diff
@@ -12,6 +12,10 @@ Dates are in `yyyy-mm-dd`.
 
 ### Fixed
 
+## Version 1.8.6 (verCode 1080600), 2021-06-16
+### Changed
+* ProgressDialog in TokenActivity with on screen prompts
+
```
Comment (abhijeetviswa) [inline CHANGELOG.md:18]: You don't need to mark a new version in this commit. We'll do that when we are ready for a release.
Instead just add an entry under Unreleased->Changed. When we are ready for a release, we'll change the Unreleased heading to the new version number.
https://github.com/crux-bphc/CMS-Android/commit/545d607c1c7457c543121fb980f1a862a416c864
Check this commit (and the immediately next child commit on the` development` branch).

File: app/build.gradle [hunk 5]
```diff
@@ -28,8 +28,8 @@ android {
         applicationId "crux.bphc.cms"
         minSdkVersion 24
         targetSdkVersion 29
-        versionCode 1080500
-        versionName "1.8.5"
+        versionCode 1080600
+        versionName "1.8.6"
```
Comment (abhijeetviswa) [inline app/build.gradle:32]: Again, you don't need to bump version in this commit. This will be done in a separate commit, which will then be tagged using `git-tag`.

File: app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt (hunk 3 above)
Comment (abhijeetviswa) [reply app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt:43]: I guess it is fine in this case since the `MutableLiveData` and `LiveData` have the same name and both of them are required, one for pushing data and for observing.

File: app/src/main/java/crux/bphc/cms/models/SingleLiveEvent.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package crux.bphc.cms.models
```
Comment (abhijeetviswa) [inline app/src/main/java/crux/bphc/cms/models/SingleLiveEvent.kt:1]: This is a utility class right? Might be better putting it on `crux.bphc.cms.utils`.

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -1,7 +1,7 @@
 <resources>
     <!-- General Strings, used in more than one place -->
     <string name="app_name">CMS BPHC</string>
-    <string name="powered_by_crux">Powered by Crux</string>
+    <string name="powered_by_crux">Powered by CRUx</string>
```
Comment (abhijeetviswa) [inline app/src/main/res/values/strings.xml:4]: Make a separate commit for this change since it's not relevant to this commit.

File: app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,223 @@
+package crux.bphc.cms.activities
```
Comment (abhijeetviswa) [inline app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt:1]: Move this viewmodel into a separate package, either under `crux.bphc.cms.activities` or within a separate `crux.bphc.cms.viewmodels`. We don't want a cluttered codebase.

File: app/src/main/res/layout/activity_token.xml [hunk 9]
```diff
@@ -1,59 +1,122 @@
...
+            app:layout_anchorGravity="center"
+            app:tint="?tokenImageTint" />
+
+        <ImageView
+            android:id="@+id/bits_logo_image"
+            android:layout_width="150dp"
+            android:layout_height="150dp"
+            android:layout_marginTop="16dp"
+            android:layout_marginBottom="8dp"
+            android:contentDescription="@string/bits_pilani_logo"
```
Comment (abhijeetviswa) [inline app/src/main/res/layout/activity_token.xml:32]: Set this to `@null`. This image is purely decorative and doesn't need to be read by a talkback reader.

Comment (abhijeetviswa) [reply app/src/main/res/layout/activity_token.xml:32]: Remove `bits_pilani_logo` in string.xml as well.

File: app/src/main/res/layout/activity_token.xml [hunk 10]
```diff
@@ -1,59 +1,122 @@
...
+            app:layout_constraintBottom_toTopOf="@+id/cms_text"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent"
+            app:layout_constraintVertical_chainStyle="packed"
+            app:srcCompat="@drawable/ic_bits_logo" />
+
+        <TextView
```
Comment (abhijeetviswa) [inline app/src/main/res/layout/activity_token.xml:40]: `android:importantForAccessibility="no"` Could probably be set for this TextView.

File: app/src/main/res/layout/activity_token.xml [hunk 11]
```diff
@@ -1,59 +1,122 @@
...
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:indeterminate="true"
+            app:layout_constraintBottom_toTopOf="@+id/login_status"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@+id/submit_button"
+            app:layout_constraintVertical_chainStyle="packed" />
+
+        <TextView
```
Comment (abhijeetviswa) [inline app/src/main/res/layout/activity_token.xml:93]: `android:importantForAccessibility="no"` can be set here as well.

File: app/src/main/res/layout/activity_token.xml [hunk 12]
```diff
@@ -1,59 +1,122 @@
...
+            android:layout_marginTop="4dp"
+            android:gravity="center"
+            android:text="@string/title_bphc"
+            android:textAppearance="@style/TextAppearance.MaterialComponents.Subtitle1"
+            app:layout_constraintBottom_toTopOf="@+id/submit_button"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@+id/cms_text" />
+
+        <TextView
```
Comment (abhijeetviswa) [inline app/src/main/res/layout/activity_token.xml:107]: `android:importantForAccessibility="no"` can be set here as well.

File: app/src/main/java/crux/bphc/cms/models/SingleLiveEvent.kt (hunk 6 above)
Comment (aryanarora180) [reply app/src/main/java/crux/bphc/cms/models/SingleLiveEvent.kt:1]: The utilities are classes with common functions that are used throughout the app. SingleLiveEvent makes sure that updated LiveData values are not handled again when the app is reopened for example. For example, for an error event, the Toast/Snackbar should be displayed only once so it checks whether it's been handled or not and only returns the value then. So it's assisting us in modelling the data which is why I kept it in `crux.bphc.cms.models.`

File: CHANGELOG.md [hunk 13]
```diff
@@ -17,6 +17,12 @@ Dates are in `yyyy-mm-dd`.
 * Incorrect theme of system navigation bar
 * No refresh indicator when opening notifications
 
+### Changed
+* ProgressDialog in TokenActivity to on screen prompts
+
+### Changed
+* ProgressDialog in TokenActivity with on screen prompts
+
```
Comment (abhijeetviswa) [inline CHANGELOG.md:25]: These should come under Unreleased.  Changes under a version number are released changes.
Read [this](https://github.com/crux-bphc/CMS-Android/blob/development/CONTRIBUTING.md) and [this](https://keepachangelog.com/en/1.0.0/) to learn more.

File: app/src/main/res/layout/activity_token.xml [hunk 14]
```diff
@@ -1,59 +1,124 @@
...
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginEnd="16dp"
+            android:layout_marginBottom="16dp"
+            android:gravity="center"
+            android:text="@string/powered_by_crux"
+            android:textAppearance="@style/TextAppearance.MaterialComponents.Headline6"
+            android:textSize="18sp"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent" />
```
Comment (abhijeetviswa) [inline app/src/main/res/layout/activity_token.xml:96]: Add `android:importantForAccessibility="no"`

**Review by abhijeetviswa — COMMENTED** [review]
Couple more minor changes and we should be good to go

File: CHANGELOG.md (hunk 13 above)
Comment (abhijeetviswa) [reply CHANGELOG.md:25]: Also, the same changelog (+ heading) has duplicated.
Also also, changelog should be written for the end user. They don't know what a `ProgressDialog` is. A more appropriate changelog is _A new and modern looking and more performance Login Screen_ or something, because at the end of the day, it is just a cosmetic change for the end user (irrespective of how relavant the technical changes might have been)

**Review by abhijeetviswa — APPROVED** [review]
Nice job. LGTM!
